crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -794,36 +793,94 @@ async fn abort_tasks<T: 'static>(mut join_set: JoinSet<T>) {
-{
+impl AwsSdkCredentialsProvider {
+    fn new(config: &aws_config::SdkConfig) -> anyhow::Result<Self> {
+        let identity_cache = config
+            .identity_cache()
+            .context("Could not find AWS credentials provider")?;
+
+        let credentials_provider = config
+            .credentials_provider()
+            .context("Could not find AWS credentials provider")?;
+
+        let identity_resolver =
+            aws_smithy_runtime_api::client::identity::SharedIdentityResolver::new(
+                credentials_provider,
+            );
+
+        // runtime_components will be ignored by the credentials provider
+        // but, we still need to create one, which is really not easy.
+        // https://github.com/awslabs/aws-sdk-rust/discussions/923#discussioncomment-7471550
+        let runtime_components = RuntimeComponentsBuilder::for_tests()
+            .with_time_source(Some(aws_smithy_async::time::SystemTimeSource::new()))
+            .with_sleep_impl(Some(aws_smithy_async::rt::sleep::TokioSleep::new()))
+            .build()
+            .context("Could not build AWS runtime components")?;
+
+        Ok(Self {
+            identity_cache,
+            identity_resolver,
+            runtime_components,
+            config_bag: aws_smithy_types::config_bag::ConfigBag::base(),
+        })
+    }
+}
+
+impl object_store::CredentialProvider for AwsSdkCredentialsProvider {
     type Credential = object_store::aws::AwsCredential;
 
-    async fn get_credential(&self) -> object_store::Result<Arc<Self::Credential>> {
-        let creds = self
-            .credentials_provider
-            .provide_credentials()
-            .await
-            .map_err(|e| {
-                // object_store's error detail rendering is not great but aws_config logs the
-                // detailed underlying cause at WARN level so we don't need to do it again here
-                object_store::Error::Unauthenticated {
+    fn get_credential<'a, 'async_trait>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn std::future::Future<Output = object_store::Result<Arc<Self::Credential>>>
+                + Send
+                + 'async_trait,
+        >,
+    >
+    where
+        Self: 'async_trait,
+        'a: 'async_trait,
+    {
+        async {
+            let identity = self
+                .identity_cache
+                .resolve_cached_identity(
+                    self.identity_resolver.clone(),
+                    &self.runtime_components,
+                    &self.config_bag,
+                )
+                .await
+                .map_err(|e| {
+                    // object_store's error detail rendering is not great but aws_config logs the
+                    // detailed underlying cause at WARN level so we don't need to do it again here
+                    object_store::Error::Unauthenticated {
+                        path: "<n/a>".to_string(),
+                        source: e.into(),
```

Comment:
Seems like we can skip `into()` here:
```suggestion
                        source: e,
```